src/bootstrap/connect.rs

```diff
@@ -0,0 +1,84 @@
+use std::sync::Arc;
+
+use http_body_util::combinators::BoxBody;
+use hyper::body::{Bytes, Incoming};
+use hyper::upgrade::Upgraded;
+use hyper::{Method, Request, Response};
+use hyper_util::rt::TokioIo;
+use tokio::net::TcpStream;
+
+use crate::empty;
+use crate::error::Error;
+
+pub(crate) fn is_connect_request(req: &Request<Incoming>) -> bool {
+    Method::CONNECT == req.method()
+}
+
+pub(crate) async fn try_upgrade(
+    req: Request<Incoming>,
+    gateway_origin: Arc<String>,
+) -> Result<Response<BoxBody<Bytes, hyper::Error>>, Error> {
+    if let Some(addr) = find_allowable_gateway(&req, &gateway_origin) {
+        tokio::task::spawn(async move {
+            match hyper::upgrade::on(req).await {
+                Ok(upgraded) => {
+                    if let Err(e) = tunnel(upgraded, addr).await {
+                        eprintln!("server io error: {}", e);
+                    };
+                }
+                Err(e) => eprintln!("upgrade error: {}", e),
+            }
```

Comment:
I don't love that these errors, especially the second, can't return HTTP responses but waiting for the update to potentially respond with an error seems to block indefinitely. It makes me feel better that this is [how the hyper proxy example does it too](https://github.com/hyperium/hyper/blob/master/examples/http_proxy.rs)